### PR TITLE
TST+BRK: How super is resolved with symlinks, needs resolution

### PR DESCRIPTION
**Pull request by mih**
Stumbled upon this question when looking for holes in the tests. Please share your thoughts.

### CODE REVIEW

**Comment by yarikoptic** [comment]
It is different so that it is a virtual Yarik adding the label. I was still sleeping then.
I think it might be the waffle.io I was trying out... Might need to disable now

**Comment by bpoldrack** [comment]
Hm. I think that means, that `GitRepo.get_toppath` should not resolve symlinks. But I'm currently not sure, whether this approach will fire back in a slightly different setup using symlinks. Any thought on this approach?

**Comment by yarikoptic** [comment]
`GitRepo.get_toppath` -- I think I have "fixed" it to try to figure out the one which would be based on PWD ... in ad55e45ce2fdfcd609103015d781a776676dd325  which is a part of Michael's #959 I believe

**Comment by mih** [comment]
But the question is: what do we want it to do? The subdataset in question is not actually underneath the super dataset.

**Comment by yarikoptic** [comment]
but what is a super-dataset of symlinked dataset? e.g. if I have

    ./super1/ds1
    ./super2/ds1 -> ../super1/ds1

and I am under ./super2/ds1 -- which one should be considered a superdataset?

if super1 -- we need to follow resolved path, if super2 -- we should follow non resolved (similarly to how I did for get_toppath).

**Comment by bpoldrack** [comment]
Well, I think if you are under `./super2/ds1` it should return `super2` and if you are under `./super1/ds1` it should retunr `super1`. What would be a situation where you possibly want a different result?

**Comment by bpoldrack** [comment]
And: As @mih pointed out in line 131 of `test_dataset.py` it yields `None` currently. (Not sure, whether this is still the case with your fix). And that means, `get_superdataset` resolved the symlink, went up and found no dataset anymore, therefore returning `None`.

**Comment by yarikoptic** [comment]
> Well, I think if you are under ./super2/ds1 it should return super2 and if you are under ./super1/ds1 it > should retunr super1. What would be a situation where you possibly want a different result?

;-) we are organizing super/sub datasets now using submodules mechanism, correct? ;)  so, if we stay true to our concept, super2 is not is not a superdataset of ds1 since there it is merely a symlink to somewhere, where it is (potentially) a proper sub-dataset.  super2 has no clue how to `get/install` ds1 (no submodule) -- it just kinda comes available if it is available in the original symlinked location.

**Comment by mih** [comment]
Are we still talking about the same thing? In my test there is a proper super/sub relationship -- that is ignored. I think this is a problem.

**Comment by mih** [comment]
And IMHO this is one case of a more general problem where we assume dataset hierarchy to be reflected on the filesystem. We don't have the same problem looking down, because we can ask the repo for it's submodules. But upwards we make inappropriate shortcuts. I think.

**Comment by yarikoptic** [comment]
> In my test there is a proper super/sub relationship
I will read the code later whenever have spare time... once again -- if only PR description had a nice description so code reading was not necessary to understand what the PR is about...  ;)

> And IMHO this is one case of a more general problem where we assume dataset hierarchy to be reflected on the filesystem. We don't have the same problem looking down, because we can ask the repo for it's submodules. But upwards we make inappropriate shortcuts. I think.

They are appropriate under assumption of a hierarchical layout.  If it is not hierarchical -- then we are indeed in trouble, and need to store the path of the super dataset e.g. within `.git/config` so we could just look up in the config.  That should then work for both (I guess).

**Comment by bpoldrack** [comment]
> super2 has no clue how to get/install ds1 (no submodule)
I think Michael is right - this is not true in his setup.

Thinking about it I realized this is indeed a more general issue. We probably don't even need to ask for it's superdataset to get in trouble, but just instantiate `Dataset` both ways. `Dataset` itself resolves symlinks and this would lead to forget where we came from, meaning both objects would be the same. Originally this was intentional but with a usecase like this, it's troublesome.

Would you guys generally agree to change `Dataset` to not resolve symlinks when instantiated and just use `normpath` instead of `realpath` in the constructor? That way we would generally be able to reflect a situation like this. That would make it harder to figure out, whether two datasets are equal (representing the same repository), but I think we hardly need such a comparison outside test code.
Any fears regarding this?

**Comment by mih** [comment]
I have no fear. Someone needs to be brave enough to make that change, discover that the tests are not complete, open a PR, and face the echo. I can do that, unless somebody beats me too it ;-)

**Comment by yarikoptic** [comment]
@bpoldrack  since we agreed awhile back we indeed should avoid `realpath` (and also `abspath` which some times leads to resolving links)

**Comment by yarikoptic** [comment]
I think that all the symlinks will put us into grave or at least cause lots of problems... and having a submodule which is also actually just a symlink outside somewhere (now that I think I have read the test correctly) is one of those "interesting" cases I would say.  I have pushed into this PR few commits -- cherry picked one I was talking about and additional where I have introduced check for super to have original one as a submodule, and then also what I thought could be correct additional condition for post-treatment of the deduced super's path... but that broke some other test ... feel free to fix more if you see clearly what is happening or scrape my commits.

But overall -- is this a use-case for flat hierarchies or ???

**Comment by mih** [comment]
This was not to promote a use case, but came from a real-life observation. I produced such a configuration for an analysis repo, because `annex init` was taking so long that I simply symlinked the source. Which worked fine for almost anything, but led to the discovery that the present behavior is unexpected and untested.

**Comment by yarikoptic** [comment]
If we are to allow such use-case, we should may be just somehow directly support it in `install` with e.g. `--beyond-reckless` ;)

File: datalad/distribution/dataset.py [hunk 1]
```diff
@@ -384,11 +393,14 @@ def get_superdataset(self, topmost=False):
             # None was found
             return None
 
-        if realpath(self.path) != self.path:
+        # and it is a subdirectory, not some fancy symlink somewhere outside
+        if realpath(self.path) != self.path and \
+            realpath(self.path).startswith(with_pathsep(realpath(sds_path))):
```
Comment (yarikoptic) [inline datalad/distribution/dataset.py]: FTR:  in https://github.com/datalad/datalad/pull/1011/files#diff-3947667888c8ff9d4bfc6c26f2be70f9L387 I overhauled this a bit -- apparently I didn't need this check here when `normpath` is used.

**Comment by mih** [comment]
Do you guys come to any conclusion on this one?

**Comment by mih** [comment]
Conclusion: no conclusion.

**Pull request closed without merging**